### PR TITLE
Help implement Metals' infer expected type feature

### PR DESCRIPTION
**Pull request by dwijnand**
Apart from helping to implement the new InferExpectedType feature in Metals, I fixed a bug in Applications, which is actually a re-implementation of my fix in #18269, except now it also applies to varargs arguments.  Previously, in #18269, I checked the typed args, but varargs specifically removes (dropInPlace) the - possibly error-typed! - var arg args, wrapping them in an not error-typed SeqLiteral tree.  Now we record those failures instantly so we don't lose them.

### CODE REVIEW

File: presentation-compiler/src/main/dotty/tools/pc/ScalaPresentationCompiler.scala [hunk 1]
```diff
@@ -198,6 +199,15 @@ case class ScalaPresentationCompiler(
         .asJava
     }
 
+  def inferExpectedType(params: OffsetParams): CompletableFuture[ju.Optional[String]] =
```
Comment (tgodzik) [inline presentation-compiler/src/main/dotty/tools/pc/ScalaPresentationCompiler.scala:202]: There is already an existing method `insertInferredType` wouldn't that be the same thing?

File: presentation-compiler/src/main/dotty/tools/pc/InferExpectedType.scala [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import dotty.tools.pc.IndexedContext
+import dotty.tools.pc.printer.ShortenedTypePrinter
+import dotty.tools.pc.printer.ShortenedTypePrinter.IncludeDefaultParam
+import dotty.tools.pc.utils.InteractiveEnrichments.*
+
+import scala.meta.internal.metals.ReportContext
+import scala.meta.pc.OffsetParams
+import scala.meta.pc.SymbolSearch
+
+class InferExpectedType(
```
Comment (tgodzik) [inline presentation-compiler/src/main/dotty/tools/pc/InferExpectedType.scala:27]: Should we instead reuse this or parts of it in InferredTypeProvider?

**Review by tgodzik — COMMENTED** [review]
Thanks for working on this :tada:

**Comment by dwijnand** [comment]
Those are pieces that @kasiaMarek setup, so I would need her help to answer, as I don't know 😄

**Comment by michelou** [comment]
@dwijnand Thanks for adding `private def instDecision(...)` in `typer/Inferencing.scala`. It makes the reasoning more clear.

File: presentation-compiler/src/main/dotty/tools/pc/ScalaPresentationCompiler.scala (hunk 1 above)
Comment (kasiaMarek) [reply presentation-compiler/src/main/dotty/tools/pc/ScalaPresentationCompiler.scala:202]: This is just for the purpose of testing. The method will be used only inside of pc by such things like completions. This covers more cases then `insertInferredType` but doesn't change it into textEdit. We could have `insertInferredType` cover those cases too but I don't think it's ever useful.

File: presentation-compiler/src/main/dotty/tools/pc/InferExpectedType.scala (hunk 2 above)
Comment (kasiaMarek) [reply presentation-compiler/src/main/dotty/tools/pc/InferExpectedType.scala:27]: It seems that `InferredTypeProvider` still processes `path` to figure out where to add edit and if/what parenthesis should be added. So I'm not sure there is any benefit in this.

**Pull request merged**